### ISSUE TITLE
synth-domain crashes dnsmasq code

### ISSUE THREAD

**Issue by cwadds** [pi-hole/FTL#367]
**In raising this issue, I confirm the following (please check boxes, eg [X]) Failure to fill the template will close your issue:**

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] The issue I am reporting can be *replicated*
- [X] The issue I am reporting isn't a duplicate 

**How familiar are you with the codebase?:**

1

---

**[BUG | ISSUE] Expected Behaviour:**
pihole-FTl running, resolving DNS queries

**[BUG | ISSUE] Actual Behaviour:**
pihole-FTL crashed, will not restart

**[BUG | ISSUE] Steps to reproduce:**
Upgrade Pi-hole using command: pihole -up
Upgrade occurred, but pihole-FTL crashes and will not restart

Rebooted Raspberry-Pi, pihole-FTL still will not start

Tried running: pihole-FTL dnsmasq-test Result: Segmentation fault

**Log file output [if available]**

[2018-08-29 08:37:59.019] ########## FTL started! ##########
[2018-08-29 08:37:59.020] FTL branch: 
[2018-08-29 08:37:59.021] FTL version: v4.0
[2018-08-29 08:37:59.021] FTL commit: 8493df4
[2018-08-29 08:37:59.021] FTL date: 2018-08-05 13:40:30 -0700
[2018-08-29 08:37:59.021] FTL user: pihole
[2018-08-29 08:37:59.023] Starting config file parsing (/etc/pihole/pihole-FTL.conf)
[2018-08-29 08:37:59.023]    SOCKET_LISTENING: only local
[2018-08-29 08:37:59.023]    AAAA_QUERY_ANALYSIS: Show AAAA queries
[2018-08-29 08:37:59.023]    MAXDBDAYS: max age for stored queries is 365 days
[2018-08-29 08:37:59.023]    RESOLVE_IPV6: Resolve IPv6 addresses
[2018-08-29 08:37:59.023]    RESOLVE_IPV4: Resolve IPv4 addresses
[2018-08-29 08:37:59.023]    DBINTERVAL: saving to DB file every minute
[2018-08-29 08:37:59.024]    DBFILE: Using /etc/pihole/pihole-FTL.db
[2018-08-29 08:37:59.024]    MAXLOGAGE: Importing up to 24.0 hours of log data
[2018-08-29 08:37:59.024]    PRIVACYLEVEL: Set to 0
[2018-08-29 08:37:59.024]    IGNORE_LOCALHOST: Show queries from localhost
[2018-08-29 08:37:59.024]    BLOCKINGMODE: Null IPs for blocked domains
[2018-08-29 08:37:59.024]    REGEX_DEBUGMODE: Inactive
[2018-08-29 08:37:59.024] Finished config file parsing
[2018-08-29 08:37:59.025] Compiled 0 Regex filters and 10 whitelisted domains in 0.8 msec (0 errors)
[2018-08-29 08:37:59.052] Database successfully initialized
[2018-08-29 08:37:59.066] Notice: Increasing queries struct size from 0 to 10000
[2018-08-29 08:37:59.067] Notice: Increasing domains struct size from 0 to 1000
[2018-08-29 08:37:59.067] Notice: Increasing clients struct size from 0 to 10
[2018-08-29 08:37:59.067] New forward server: 103.86.99.100 (0/0)
[2018-08-29 08:37:59.067] Notice: Increasing forwarded struct size from 0 to 4
[2018-08-29 08:37:59.067] Notice: Increasing overTime struct size from 0 to 100
[2018-08-29 08:37:59.067] New forward server: 103.86.96.100 (1/4)
[2018-08-29 08:37:59.169] Notice: Increasing queries struct size from 10000 to 20000
[2018-08-29 08:37:59.264] Notice: Increasing queries struct size from 20000 to 30000
[2018-08-29 08:37:59.358] Notice: Increasing queries struct size from 30000 to 40000
[2018-08-29 08:37:59.361] Notice: Increasing clients struct size from 10 to 20
[2018-08-29 08:37:59.453] Notice: Increasing queries struct size from 40000 to 50000
[2018-08-29 08:37:59.524] Notice: Increasing overTime struct size from 100 to 200
[2018-08-29 08:37:59.543] Imported 48961 queries from the long-term database
[2018-08-29 08:37:59.544]  -> Total DNS queries: 48961
[2018-08-29 08:37:59.544]  -> Cached DNS queries: 1112
[2018-08-29 08:37:59.544]  -> Forwarded DNS queries: 2841
[2018-08-29 08:37:59.544]  -> Exactly blocked DNS queries: 1214
[2018-08-29 08:37:59.544]  -> Unknown DNS queries: 43794
[2018-08-29 08:37:59.544]  -> Unique domains: 287
[2018-08-29 08:37:59.545]  -> Unique clients: 15
[2018-08-29 08:37:59.545]  -> Known forward destinations: 2
[2018-08-29 08:37:59.545] Successfully accessed setupVars.conf
[2018-08-29 08:37:59.554] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[2018-08-29 08:37:59.554] ---------------------------->  FTL crashed!  <----------------------------
[2018-08-29 08:37:59.554] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[2018-08-29 08:37:59.554] Please report a bug at https://github.com/pi-hole/FTL/issues
[2018-08-29 08:37:59.554] and include in your report already the following details:

[2018-08-29 08:37:59.554] FTL has been running for 0 seconds
[2018-08-29 08:37:59.554] FTL branch: 
[2018-08-29 08:37:59.554] FTL version: v4.0
[2018-08-29 08:37:59.554] FTL commit: 8493df4
[2018-08-29 08:37:59.554] FTL date: 2018-08-05 13:40:30 -0700
[2018-08-29 08:37:59.555] FTL user: pihole
[2018-08-29 08:37:59.555] Received signal: Segmentation fault
[2018-08-29 08:37:59.555]      at address: 0
[2018-08-29 08:37:59.555]      with code: SEGV_MAPERR (Address not mapped to object)
[2018-08-29 08:37:59.555] Backtrace:
[2018-08-29 08:37:59.555] B[0000]: /usr/bin/pihole-FTL(+0x1870a) [0x4af70a]
[2018-08-29 08:37:59.556] B[0001]: /lib/arm-linux-gnueabihf/libc.so.6(__default_rt_sa_restorer+0) [0x76e446c0]
[2018-08-29 08:37:59.556] Thank you for helping us to improve our FTL engine!
[2018-08-29 08:37:59.556] FTL terminated!

**Device specifics**

Hardware Type: Raspberry Pi 2 Model B Rev 1.1
OS: Linux pihole 4.14.34-v7+ #1110 SMP Mon Apr 16 15:18:51 BST 2018 armv7l GNU/Linux

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

**Comment by cwadds** [comment]
Further information:
Output of command: file $(which pihole-FTL)
/usr/bin/pihole-FTL: ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-armhf.so.3, for GNU/Linux 3.2.0, BuildID[sha1]=5864a2037d20fbf1a54d3ee377c9ebbc0b7400da, not stripped

**Comment by DL6ER** [comment]
Please run
```
addr2line 0x1870a -e $(which pihole-FTL)
addr2line 0x4af70a -e $(which pihole-FTL)
```

**Comment by cwadds** [comment]
Hi,Results of commands below:root@pihole:~# addr2line 0x1870a -e $(which pihole-FTL)/root/project/signals.c:46root@pihole:~# addr2line 0x4af70a -e $(which pihole-FTL)??:0root@pihole:~#On 29 August 2018 at 17:02 DL6ER <notifications@github.com> wrote:Please runaddr2line 0x1870a -e $(which pihole-FTL)
addr2line 0x4af70a -e $(which pihole-FTL)

**Comment by DL6ER** [comment]
Okay, thanks. Unfortunately, that isn't very helpful.

Could you run
```
sudo service pihole stop
sudo gdb -ex run --args $(which pihole-FTL) debug
```
?

When it crashes, copy the output here. Please also enter `backtrace` and post the output here.

**Comment by DL6ER** [comment]
`pihole-FTL` crashes while parsing the file `/etc/dnsmasq.d/03-reverse.conf`. I don't recall it being a standard Pi-hole provided file. What are its contents?

From the backtrace, I *assume* that the failure is triggered by a `synth-domain=` (or domain=` ?) line in this file.

**Comment by DL6ER** [comment]
@cwadds Looking again at the `synth-domain` code part of `dnsmasq`, it seems quite likely that the variable `new->prefix` may be used uninitialized but I can only be sure once I can reproduce this crash using your config file.

**Comment by cwadds** [comment]
Good Morning (Australia time),You are correct, the contents of the reverse lookup file is as follows:synth-domain=wadds.net.au,192.168.200.0/24Firstly, I removed the offending file and tried to restart pihole without success.I then rebooted the Pi and all is running as expected.Thanks for your assistance.Is there any likelihood of being able to use the reverse lookup functionality in pihole?It is not a critical function for me, but I have found it useful. :-)Cheers,ConradOn 30 August 2018 at 20:02 DL6ER <notifications@github.com> wrote:pihole-FTL crashes while parsing the file /etc/dnsmasq.d/03-reverse.conf. I don't recall it being a standard Pi-hole provided file. What are its contents?From the backtrace, I assume that the failure is triggered by a synth-domain= (or domain=` ?) line in this file.—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.  Conrad WaddsRed Hat EngineerABN: 94 074 170 101PO Box 655 Gosford NSW 2250Phone: 0418 430 015 | CWadds@wadds.net.au

**Comment by DL6ER** [comment]
> Is there any likelihood of being able to use the reverse lookup functionality in pihole?

Yes, this is a bug (we adopted this bug from `dnsmasq 2.79`) and will be fixed. The prefix is, unfortunately, not as optional as the documentation suggests.

You could use
```
synth-domain=wadds.net.au,192.168.200.0/24,internal
```
to obtain
```
internal-192-168-200-1.wadds.net.au
```

I'll work on a bug fix. Thanks for reporting this!

**Comment by cwadds** [comment]
Thanks for your efforts.

I'll try your suggestion.

**Comment by DL6ER** [comment]
This bug is a confirmed `dnsmasq` bug introduced in version 2.79 (which happens to be the version we currently bundle with `pihole-FTL`). This bug will be fixed by #369 for the next release Pi-hole v4.1

**Comment by DL6ER** [comment]
Note: This issue will be auto-closed once Pi-hole FTL v4.1 is released
